staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package responsewriter
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+)
+
+// UserProvidedDecorator represensts a user (client that uses this package)
+// provided decorator that composes an inner http.ResponseWriter object.
+// The user provided decorator object must return the inner (decorated)
+// http.ResponseWriter object via the Unwrap function.
+type UserProvidedDecorator interface {
+	http.ResponseWriter
+
+	// Unwrap returns the inner http.ResponseWriter object associated
+	// with the user provided decorator.
+	Unwrap() http.ResponseWriter
+}
+
+// Wrap accepts a user provided decorator of an "inner" http.responseWriter
+// object and potentially wraps it with a new http.ResponseWriter object that
+// implements http.CloseNotifier, http.Flusher and/or http.Hijacker
+// accordingly by delegating to the user provided decorator or the inner
+// http.ResponseWriter, so that the returned http.ResponseWriter object
+// implements the same subset of those interfaces as the inner http.ResponseWriter.
```

Comment:
This function correctly implements just 3 of the 8 possible cases.  This comment should enumerate those three cases.

Also, for the sake of future-proofing, it might be best if the function name clearly indicated that it is covering just those three particular cases.  Spelling three cases out in words will be a bit much.  Coding the extension interfaces as bits makes each case a small number.  We could name this function "Wrap037", identifying the three covered cases, for example.

Reply:
This has been resolved.